default/python/AI/SaveGameManager.py

```diff
@@ -0,0 +1,260 @@
+        elif isinstance(o, dict):
+            return self._dict_encoder(o)
+        elif isinstance(o, list):
+            return self._list_encoder(o)
+        elif isinstance(o, tuple):
+            o = self._tuple_encoder(o)
+        elif isinstance(o, set):
+            o = self._set_encoder(o)
+        else:
+            return self._object_encoder(o)
+        return "\"%s\"" % o
+
+    def _object_encoder(self, obj):
+        try:
+            if obj.__class__.__module__ not in _trusted_modules:
+                raise CanNotSaveGameException("Module %s is not trusted" % obj.__class__.__module__)
+        except AttributeError:
+            # obj does not have a class or class has no module
+            raise CanNotSaveGameException("Encountered unsupported object %s (%s)" % (obj, type(obj)))
+
+        key = CLASS_PREFIX + obj.__class__.__module__ + '.' + obj.__class__.__name__
+
+        try:
+            getstate = obj.__getstate__
+        except AttributeError:
+            value = obj.__dict__
+        else:
+            # only call now to not incorrectly catch exceptions during getstate call
+            value = getstate()
+        return self._dict_encoder({key: value})
+
+    def _list_encoder(self, o):
+        return '[' + ', '.join(self.encode(v) for v in o) + ']'
+
+    def _tuple_encoder(self, o):
+        return (TUPLE_PREFIX + "(" + self._list_encoder(list(o)) + ")").replace('\"', PLACEHOLDER)
+
+    def _set_encoder(self, o):
+        return (SET_PREFIX + "(" + self._list_encoder(list(o)) + ")").replace('\"', PLACEHOLDER)
+
+    def _dict_encoder(self, o):
+        return '{' + ', '.join('%s: %s' % (self.encode(k), self.encode(v)) for k, v in o.iteritems()) + '}'
+
+
+class FreeOrionAISaveGameDecoder(json.JSONDecoder):
+
+    def __init__(self, **kwargs):
+        super(FreeOrionAISaveGameDecoder, self).__init__(object_hook=self.__parse_savegame, strict=False, **kwargs)
+
+    def decode(self, s, _w=None):
+        return self.__interpret(super(FreeOrionAISaveGameDecoder, self).decode(s))
+
+    def __parse_savegame(self, obj):
+        """Recursively parse a savegame, loading only trusted classes."""
+        # print "Parsing: ", type(obj), obj
+        if not isinstance(obj, dict):
+            retval = self.__interpret(obj)
+            # print "Returns: ", retval
+            return retval
+
+        if len(obj) != 1:
+            # by convention, this must be a standard dict because our custom encoded classes
+            # have only a single key # which is its name and its __dict__ content as value.
+            retval = self.__interpret(obj)
+            # print "Returns: ", retval
+            return retval
+
+        [(key, value)] = obj.items()
+        if key.startswith(CLASS_PREFIX):
+            # get the module and class name
+            full_name = key[len(CLASS_PREFIX):]
+            module_name, class_name = full_name.rsplit('.', 1)
+            if module_name not in _trusted_modules:
```

Comment:
Seems like it would be better to test the class name rather than the module name...